core/server/worker/src/main/java/alluxio/worker/netty/UfsFileWriteHandler.java

```diff
@@ -0,0 +1,226 @@
+      mOutputStream = outputStream;
+    }
+
+    /**
+     * @param underFileSystem UFS to set
+     */
+    public void setUnderFileSystem(UnderFileSystem underFileSystem) {
+      mUnderFileSystem = underFileSystem;
+    }
+
+    /**
+     * @param counter counter to set
+     */
+    public void setCounter(Counter counter) {
+      mCounter = counter;
+    }
+  }
+
+  /**
+   * Creates an instance of {@link UfsFileWriteHandler}.
+   *
+   * @param executorService the executor service to run {@link PacketWriter}s
+   * @param ufsManager the file data manager
+   */
+  UfsFileWriteHandler(ExecutorService executorService, UfsManager ufsManager) {
+    super(executorService);
+    mUfsManager = ufsManager;
+  }
+
+  @Override
+  protected boolean acceptMessage(Object object) {
+    if (!super.acceptMessage(object)) {
+      return false;
+    }
+    Protocol.WriteRequest request = ((RPCProtoMessage) object).getMessage().asWriteRequest();
+    return request.getType() == Protocol.RequestType.UFS_FILE;
+  }
+
+  @Override
+  protected UfsFileWriteRequest createRequest(RPCProtoMessage msg) throws Exception {
+    return new UfsFileWriteRequest(msg.getMessage().asWriteRequest());
+  }
+
+  @Override
+  protected void completeRequest(Channel channel) throws Exception {
+    UfsFileWriteRequest request = getRequest();
+    if (request == null) {
+      return;
+    }
+    if (request.getOutputStream() == null) {
+      createUfsFile(channel);
+    }
+    Preconditions.checkState(request.getOutputStream() != null);
+    request.getOutputStream().close();
+    request.setOutputStream(null);
+  }
+
+  @Override
+  protected void cancelRequest() throws Exception {
+    UfsFileWriteRequest request = getRequest();
+    if (request == null) {
+      return;
+    }
+    // TODO(calvin): Consider adding cancel to the ufs stream api.
+    if (request.getOutputStream() != null
+        && request.getUnderFileSystem() != null) {
+      request.getOutputStream().close();
+      request.getUnderFileSystem().deleteFile(request.getUfsPath());
+      request.setOutputStream(null);
+    }
+  }
+
+  @Override
+  protected void cleanupRequest() throws Exception {
+    cancelRequest();
+  }
+
+  @Override
+  protected void writeBuf(Channel channel, ByteBuf buf, long pos) throws Exception {
+    UfsFileWriteRequest request = getRequest();
+    Preconditions.checkState(request != null);
+    if (request.getOutputStream() == null) {
+      createUfsFile(channel);
+    }
+
+    buf.readBytes(getRequest().getOutputStream(), buf.readableBytes());
+  }
+
+  private void createUfsFile(Channel channel) throws IOException {
+    UfsFileWriteRequest request = getRequest();
+    Preconditions.checkState(request != null);
+    Protocol.CreateUfsFileOptions createUfsFileOptions = request.getCreateUfsFileOptions();
+    UfsInfo ufsInfo = mUfsManager.get(createUfsFileOptions.getMountId());
+    UnderFileSystem ufs = ufsInfo.getUfs();
+    request.setUnderFileSystem(ufs);
```

Comment:
If the UFS creation is not expensive would it be worth doing it when the request is created?